File: Java-Src/Layer.java
import java.util.*;

public class Layer
{
    private double _learningRate;
    private String _layerName;
    private ArrayList<Node> _layerNodes;

    public Layer(double learnRate, String name)
    {
        this._learningRate = learnRate;
        this._layerName = name;
        this._layerNodes = new ArrayList<Node>();
    }

    public double getLearningRate()
    {
        return this._learningRate;
    }

    public String getLayerName()
    {
        return this._layerName;
    }

    public ArrayList<Node> getLayerNodes()
    {
        return this._layerNodes;
    }

    public void forward()
    {
        for (Node node : this.getLayerNodes())
            node.calculate();
    }

    public void backward()
    {
        for (Node node : this.getLayerNodes())
            node.derive(this.getLearningRate());
    }

    public void addNode(Node targetNode)
    {
        this.getLayerNodes().add(targetNode);
    }

    public String toString()
    {
        return this.getLayerName();
    }
}

File: Cpp-Src/layer.cpp
#include "egghead_comp.hpp"

Layer::Layer(double learn_rate, string name)
{
    learningRate = &learn_rate;
    layerName = &name;
    layerNodes = new vector<Node>();
}

Layer::~Layer()
{
    delete learningRate;
    delete layerName;
    delete layerNodes;
}

void Layer::forward()
{
    for (int index = 0; index < this->getLayerNodes().size(); index++)
        this->getLayerNodes()[index].calculate();
}

void Layer::backward()
{
    for (int index = 0; index < this->getLayerNodes().size(); index++)
        this->getLayerNodes()[index].derive(this->getLearningRate());
}

void Layer::addNode(Node targetNode)
{
    getLayerNodes().push_back(targetNode);
}

string Layer::toString()
{
    return getLayerName();
}

double Layer::getLearningRate()
{
    return *learningRate;
}

string Layer::getLayerName()
{
    return *layerName;
}

vector<Node> Layer::getLayerNodes()
{
    return *layerNodes;
}
